### ISSUE TITLE
Bidirectional array compatibility

### ISSUE THREAD

**Issue by jonas-eschle** [jax-ml/jax#6222]
I've found a similar issue which however does not address this fully, https://github.com/google/jax/issues/1100.

I was wondering what the plans are about bidirectional conversions from other libraries than Numpy. So simply, TensorFlow can use a jax ndarray but jax can't use the TF ndarray. This would correspond to the current Numpy behavior as numpy takes the TensorFlow ndarrays.

I am not precisely sure why this limitation is there, is that since it may won't yet be able to share the memory as in https://github.com/google/jax/issues/1100?

jax checks whether the array in as ndarray but itself does not inherit from it (seems cynical :) ). Why don't we:
- use ducktyping and do a np.asarray on the object (if it ain't already jax?) or just nothing?
- allow to register array-like types (with conversion?)

I assume I am completely missing out some technical details why it ain't that trivial? What are the overall plans/roadmaps for this

The usecase is simple: it removes the dependency on one specific backend and allows to transfer slowly to another one/ be mixed.

**Comment by mayou36** [comment]
Ah I see, that's pretty close. And as I read, the preferences is to fail loud instead of silently convert, correct? That would pretty much answer my question the.

**Comment by jakevdp** [comment]
Yes, JAX has handling of numpy array inputs built-in. Numpy can handle JAX arrays because JAX arrays define the `__array__` method, which NumPy uses to allow third-party objects to specify how they should be handled by numpy, without numpy itself having to explicitly handle those third-party objects.

JAX's `__jax_array__` hook is meant to be similar to numpy's `__array__`: i.e. a method for third-party libraries to specify how they should be handled by JAX, without JAX having to explicitly handle those objects.

**Comment by hawkinsp** [comment]
I think the modern answer to this would be "use dlpack to exchange arrays between frameworks". Closing because I'm not sure there's anything for us to do here.
